File: graph_input_problems/degree_euler_graph_step4/swift/step4.swift
/*
これはpaizaラーニングの「グラフ構造の入力メニュー」-「準オイラーグラフ・無向グラフ」に
swiftでチャレンジしたコードです。
https://paiza.jp/works/mondai/graph_input_problems/graph_input_problems__degree_euler_graph_step4

作成環境
Ubuntu 20.04.2 LTS
Swift version 5.4 (swift-5.4-RELEASE) Target: x86_64-unknown-linux-gnu
*/
import Foundation

// メイン関数
func main() {
    // 1 行目に、頂点の個数を表す整数 n と、頂点の組の個数を表す整数 m が半角スペース区切りで与えられます。
    let items = readLine()!.trimmingCharacters(in: .whitespacesAndNewlines).split(separator: " ")
    let n = Int(items[0])!
    let m = Int(items[1])!
    // 続く m 行では、頂点の組 a_i, b_i が半角スペース区切りで与えられます。(1 ≦ i ≦ m)
    var count: [Int] = [Int](repeating: 0, count: n)
    for _ in 0..<m {
        let itemsG = readLine()!.trimmingCharacters(in: .whitespacesAndNewlines).split(separator: " ")
        let a = Int(itemsG[0])!
        let b = Int(itemsG[1])!
        // 頂点につながる辺をカウント
        count[a-1] += 1
        count[b-1] += 1
    }
    // すべての辺を一筆書きすることができるかどうか判定
    var odd = 0
    for v in count {
        if v % 2 == 1 {
            odd += 1
        }
    }
    // 結果出力
    if odd == 0 || odd == 2 {
        // 一筆書きできる
        print("1")
    } else {
        print("0")
    }
}

// エントリーポイント
main()

File: graph_input_problems/degree_euler_graph_step4/golang/step4.go
/*
これはpaizaラーニングの「グラフ構造の入力メニュー」-「準オイラーグラフ・無向グラフ」に
go言語でチャレンジしたコードです。
https://paiza.jp/works/mondai/graph_input_problems/graph_input_problems__degree_euler_graph_step4

作成環境
Ubuntu 20.04.2 LTS
go version go1.16.4 linux/amd64
*/

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// main はエントリーポイント
func main() {
	var sc *bufio.Scanner = bufio.NewScanner(os.Stdin)
	// 1 行目に、頂点の個数を表す整数 n と、頂点の組の個数を表す整数 m が半角スペース区切りで与えられます。
	sc.Scan()
	var items []string = strings.Split(strings.TrimSpace(sc.Text()), " ")
	var n, _ = strconv.Atoi(items[0])
	var m, _ = strconv.Atoi(items[1])
	// 続く m 行では、頂点の組 a_i, b_i が半角スペース区切りで与えられます。(1 ≦ i ≦ m)
	var degree = make([]int, n)
	for i := 0; i < m; i++ {
		sc.Scan()
		var items []string = strings.Split(strings.TrimSpace(sc.Text()), " ")
		var a, _ = strconv.Atoi(items[0])
		var b, _ = strconv.Atoi(items[1])
		// 頂点につながる辺をカウント
		degree[a-1]++
		degree[b-1]++
	}
	//連結な無向グラフにおいて、すべての辺を一筆書きすることができるための必要十分条件は
	// 以下のようになります。
	// ・次数が奇数である頂点の個数が、ちょうど 0 または 2 である。
	var odd int = 0
	for _, v := range degree {
		if v%2 != 0 {
			odd += 1
		}
	}
	var answer int
	if odd == 0 || odd == 2 {
		answer = 1
	} else {
		answer = 0
	}
	// 答えを出力する
	fmt.Println(answer)
	return
}
